httomo/methods.py

```diff
@@ -62,18 +120,71 @@ def _save_dataset_data(
     assert stop[1] <= dataset.shape[1]
     assert stop[2] <= dataset.shape[2]
     assert dataset.shape == global_shape
+    if isinstance(dataset, h5py.Dataset) and httomo.globals.COMPRESS_INTERMEDIATE:
+        # Write operations must be collective when applying compression, see
+        # https://github.com/h5py/h5py/issues/1564
+        with dataset.collective:
+            dataset[start[0] : stop[0], start[1] : stop[1], start[2] : stop[2]] = data
+        return
+
     dataset[start[0] : stop[0], start[1] : stop[1], start[2] : stop[2]] = data
 
 
-def _save_auxiliary_data(
-    file: h5py.File, angles: np.ndarray, detector_x: int, detector_y: int
+def _save_auxiliary_data_hdf5(
+    file: h5py.File,
+    angles: np.ndarray,
+    detector_x: int,
+    detector_y: int,
 ):
     # only save if not there yet
     if "/angles" in file:
         return
 
-    file.create_dataset("/angles", data=angles)
+    file.create_dataset("angles", data=angles)
+
     file_name = pathlib.Path(file.filename).name
     file.create_dataset(file_name, data=[0, 0])
     g1 = file.create_group("data_dims")
     g1.create_dataset("detector_x_y", data=[detector_x, detector_y])
+
+
+def _save_auxiliary_data_zarr(
+    store: zarr.DirectoryStore,
+    angles: np.ndarray,
+    detector_x: int,
+    detector_y: int,
+):
+    if mpiutil.comm.rank != 0:
```

Comment:
@ptim0626, @yousefmoazzam  Any reason why this condition is required for Zarr specifically? Do we need to do the auxiliary data saving for the 0th rank only for the hdf5 file as well?

Reply:
HDF5 opens the file in MPI mode (aware of the other processes), and only creates the dataset if not already present in the file....